### ISSUE TITLE
(aws-cdk/pipelines): Upload to S3 failed with the following error:  Access Denied

### ISSUE THREAD

**Issue by Stf-F** [aws/aws-cdk#13733]
### The Question
For the past 5 days I have tried to deploy a CDK Pipeline using the instructions available [here](https://docs.aws.amazon.com/cdk/latest/guide/cdk_pipeline.html) but no luck. The buckets provisioned seem to consistently be lacking the relevant policies. As soon as I deploy the pipeline I get `Upload to S3 failed with the following error: Access Denied (Service: Amazon S3; Status Code: 403; Error Code: AccessDenied; Request ID: xxxx; S3 Extended Request ID: xxxx; Proxy: null) (Service: null; Status Code: 0; Error Code: null; Request ID: null; S3 Extended Request ID: null; Proxy: null) (Service: null; Status Code: 0; Error Code: null; Request ID: null; S3 Extended Request ID: null; Proxy: null)`. I have deleted the Stacks (including the CDK Toolkit one) and restarted the whole configuration from scratch 4 or 5 times. I have tried with  the version `1.94.0` and `1.94.1` of the CDK but no luck there either.
I have partly resolved some of the issues by manually adding some bucket policies but that, in turn, only pushed the issue further down the pipeline. (First on the `UpdatePipeline` stage and then on the `Prepare` stage), where I ended up being completely stuck.

Any help would be greatly appreciated, I am running out of ideas.

Thanks

### Environment

  - **CDK CLI Version:** 1.89.0 (build df7253c)
  - **Module Version:** 1.94.0 && 1.94.1
  - **Node.js Version:** 14.15.1
  - **OS:** macOS Catalina 10.15.7 
  - **Language (Version):**  TypeScript (4.2.3)

### Other information
Deps:
```json
"devDependencies": {
    "@aws-cdk/assert": "^1.94.0",
    "@types/jest": "^26.0.21",
    "@types/node": "14.14.35",
    "aws-cdk": "^1.94.0",
    "jest": "^26.4.2",
    "ts-jest": "^26.5.4",
    "ts-node": "^9.0.0",
    "typescript": "4.2.3"
  },
  "dependencies": {
    "@aws-cdk/aws-appsync": "^1.94.0",
    "@aws-cdk/aws-cloudfront": "^1.94.0",
    "@aws-cdk/aws-cloudfront-origins": "^1.94.0",
    "@aws-cdk/aws-codebuild": "^1.94.0",
    "@aws-cdk/aws-codepipeline": "^1.94.0",
    "@aws-cdk/aws-codepipeline-actions": "^1.94.0",
    "@aws-cdk/aws-cognito": "^1.94.0",
    "@aws-cdk/aws-dynamodb": "^1.94.0",
    "@aws-cdk/aws-iam": "^1.94.0",
    "@aws-cdk/aws-kms": "^1.94.0",
    "@aws-cdk/aws-lambda": "^1.94.0",
    "@aws-cdk/aws-lambda-nodejs": "^1.94.0",
    "@aws-cdk/aws-pinpoint": "^1.94.0",
    "@aws-cdk/aws-s3": "^1.94.0",
    "@aws-cdk/aws-s3-deployment": "^1.94.0",
    "@aws-cdk/core": "^1.94.0",
    "@aws-cdk/custom-resources": "^1.94.0",
    "@aws-cdk/pipelines": "^1.94.0",
    "@aws-sdk/s3-request-presigner": "^3.9.0",
    "source-map-support": "^0.5.16"
  }
```

![image](https://user-images.githubusercontent.com/52630104/112008162-240bd600-8b1d-11eb-9354-e5d213476229.png)

Edit: Upgraded to CLI `1.94.1` and deps to `1.94.1`, deleted stacks, redeployed, same issue. The buckets are created without policies.

Edit2:  Downgraded to `"typescript": "~3.9.7"` and to  `"@types/node": "10.17.27"` to align with what ships today on a `cdk init`, deleted package lock, deleted stacks, boostrapped, redeployed and same problem.

Edit3: Since I am using AWS SSO and am using [this Python package](https://pypi.org/project/aws2-wrap/) to use SSO from the CLI I created 2 IAM users with AdminAccess on my pipeline account and my target account to rule out the IAM problem.  I deleted my package lock, deleted stacks, boostrapped, redeployed and same problem.

**Comment by rix0rrr** [comment]
> The buckets provisioned seem to consistently be lacking the relevant policies.

Which buckets specifically? The toolkit buckets or the pipeline buckets?

Is this in-account? Cross-account? Do you have code samples to share?


```
    "@aws-cdk/aws-appsync": "^1.94.0",
    "@aws-cdk/aws-cloudfront": "^1.94.0",
```

This is a problem. You MUST NOT use `^` in the version specifiers for CDK packages. This is unlikely to be causing your current issue, but it could be.

**Comment by rix0rrr** [comment]
All in all you're not giving us a lot to go on here. 

Does the CloudFormation templates you deploy have `AWS::S3::BucketPolicy` objects in them, yet they are not being reflected in the deployed infrastructure?

If so, what does the CloudFormation Event Log for the stack say?

Can you think of any account-wide configuration that could prevent bucket policies from being applied?

**Comment by kmonihen** [comment]
With limited info, a guess is that it could be the issue described here https://github.com/aws/aws-cdk/issues/13557 and fixed in 1.95 (https://github.com/aws/aws-cdk/pull/13637). Could you try with 1.95+ or as a temporary workaround set "@aws-cdk/aws-s3:grantWriteWithoutAcl": "false".

**Comment by Stf-F** [comment]
Hi @rix0rrr , I have a cross account configuration with a Github v2 integration and I am using the `BitBucketSourceAction` in my CDK code. 
Looks like the issue from the permissions given to the codestar role are not sufficient. 
This is what is generated by default upon deploy:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": "codestar-connections:UseConnection",
            "Resource": "arn:aws:codestar-connections:eu-west-1:xxxxx:connection/xxxx",xx
            "Effect": "Allow"
        },
        {
            "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:Abort*"
            ],
            "Resource": [
                "arn:aws:s3:::bucket",
                "arn:aws:s3:::bucket/*"
            ],
            "Effect": "Allow"
        },
        {
            "Action": [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*"
            ],
            "Resource": "arn:aws:kms:eu-west-1:xxxxxxx:key/xxxxx",
            "Effect": "Allow"
        }
    ]
}
```

Yet, this does not seem to be sufficient to access the artifacts buckets, which blocked the pipeline from the start in my case. By temporarily updating the actions and replacing them with `s3:*` the issue highlighted above was resolved. 
@kmonihen , just checked the issue, and yes, exactly, same problem.

Thanks

**Comment by skinny85** [comment]
Hey @Stf-F ,

this is probably related to the bug fixed in https://github.com/aws/aws-cdk/pull/13637. Make sure to use the version that contains that fix (`1.95.1` or later), and this error should go away.

Thanks,
Adam
